### DISCUSSION TITLE
Contents of an ev extras package

### DISCUSSION THREAD

**Discussion by yumaikas** [janet-lang/janet#615] [Ideas]
So, I'm noodling around the idea of an ev extras package. Right now, ev is in a place where it has some very powerful primitives for event loops and async process monitoring, but using those primitives can take a bit of work, as I found out writing some code that wraps the `net` and `sc` windows executables.

So I've drafted out some function signatures for some of the things I'm missing here https://gist.github.com/yumaikas/3b80e13f94ce92c37b38581438f0630c

For me, primary sources of inspiration include C#'s async/await and Tasks, as well as Go's channels, and the little bit of OTP I have in my head. Notable APIs that come to mind are things like C#'s [WhenAny](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenany?view=net-5.0), or the idea of auto-restarting ev fibers when they fail, akin to Erlang supervisors.

**Comment by andrewchambers** [comment]
I don't think you need the ev pipes - the stdin/stdout parts of a pipeline has no need to send data back to the janet process. Only the start of the pipeline and potentially the end need to be connected to janet ev.

The only use would perhaps be interleaving janet functions in between processes.
